rust/connlib/clients/android/src/lib.rs

```diff
@@ -195,27 +195,30 @@ impl Callbacks for CallbackHandler {
         })
     }
 
-    fn on_add_route(&self, route: IpNetwork) -> Result<(), Self::Error> {
+    fn on_add_route(&self, route: IpNetwork) -> Result<RawFd, Self::Error> {
         self.env(|mut env| {
-            let route = env.new_string(route.to_string()).map_err(|source| {
-                CallbackError::NewStringFailed {
-                    name: "route",
+            let ip = env
+                .new_string(route.network_address().to_string())
+                .map_err(|source| CallbackError::NewStringFailed {
+                    name: "route_ip",
                     source,
-                }
-            })?;
-            call_method(
-                &mut env,
+                })?;
+
+            let name = "onAddRoute";
+            env.call_method(
                 &self.callback_handler,
-                "onAddRoute",
-                "(Ljava/lang/String;)V",
-                &[JValue::from(&route)],
+                name,
+                "(Ljava/lang/String;I)I",
+                &[JValue::from(&ip), JValue::Int(route.netmask().into())],
             )
+            .and_then(|val| val.i())
+            .map_err(|source| CallbackError::CallMethodFailed { name, source })
```

Comment:
If you do `map_err` first, you can use `?` and avoid the use of the `and_then` combinator in favor of just linear function execution.

Reply:
After calling `i()` we still want to map that error to `CallMethodFailed` since we don't implement `From` `jni`'s `Error` (it needs more context information so we can't trivially implement it), so it'd be quite more verbose if we don't use `and_then` having to call `map_err` twice.

Reply:
Ah damn.